src/object/tcb.c

```diff
@@ -78,6 +78,39 @@ static inline void removeFromBitmap(word_t cpu, word_t dom, word_t prio)
     }
 }
 
+tcb_queue_t tcb_queue_remove(tcb_queue_t queue, tcb_t *tcb)
+{
+    tcb_t *before;
+    tcb_t *after;
+
+    before = tcb->tcbSchedPrev;
+    after = tcb->tcbSchedNext;
+
+    if (queue.head == tcb && queue.end == tcb) {
+        queue.head = NULL;
+        queue.end = NULL;
+    } else {
+        if (queue.head == tcb) {
+            after->tcbSchedPrev = NULL;
+            tcb->tcbSchedNext = NULL;
+            queue.head = after;
+        } else {
+            if (queue.end == tcb) {
+                before->tcbSchedNext = NULL;
+                tcb->tcbSchedPrev = NULL;
+                queue.end = before;
+            } else {
+                before->tcbSchedNext = after;
+                after->tcbSchedPrev = before;
+                tcb->tcbSchedPrev = NULL;
+                tcb->tcbSchedNext = NULL;
+            }
+        }
+    }
```

Comment:
I suspect @michaelmcinerney picked the order of comparisons specifically to make the proofs simpler, but also to avoid partial updates of the `queue_t` variable, because that gives an inconsistent view at what list it is at that point (i.e. no valid abstraction exists until the rest of the updates are done). @michaelmcinerney can you confirm?

Reply:
My initial reason for this was to reduce the unnecessary setting of the `tcbSchedPrev` and `tcbSchedNext` fields of `tcb` to `NULL` in the cases where the queue is either a singleton or `tcb` is the head or the end. The verification work for this function is done as is, so our preference would be to leave this function